File: src/cadreur/millumin.py
"""Millumin OSC I/O: UDP client out, feedback listener in, query-with-timeout.

Out (to millumin.host:port, default 127.0.0.1:5000):
    /layer:NAME/scale <f>              uniform multiplier, 1.0 = 100 %
    /layer:NAME/position/xy <f f>      pixels, canvas top-left origin

In (feedback listener on UDP feedback_port, default 8000 — must be enabled
manually in Millumin: Device manager -> OSC -> API feedback -> 127.0.0.1:8000):
    send /layer:NAME/scale/?        -> /millumin/layer:NAME/scale <f>
    send /layer:NAME/position/xy/?  -> /millumin/layer:NAME/position/xy <f f>
                                       OR /position/x <f> + /position/y <f>
Both position reply arities are tolerated (verify the real shape on day 1).
Queries correlate strictly by expected reply address; unsolicited feedback
traffic is ignored. query() BLOCKS up to the timeout — call it off the event
loop (asyncio.to_thread).
"""
from __future__ import annotations

import logging
import threading
import time
from typing import Optional

from pythonosc.dispatcher import Dispatcher
from pythonosc.osc_server import ThreadingOSCUDPServer
from pythonosc.udp_client import SimpleUDPClient

from .config import MilluminCfg

log = logging.getLogger("cadreur.millumin")


class MilluminIO:
    def __init__(self, cfg: MilluminCfg) -> None:
        self.cfg = cfg
        self._client = SimpleUDPClient(cfg.host, cfg.port)
        self._lock = threading.Lock()
        self._pending: dict[str, dict] = {}  # reply address -> query slot
        self._server: Optional[ThreadingOSCUDPServer] = None
        self._server_thread: Optional[threading.Thread] = None

    # --- lifecycle ------------------------------------------------------------
    def start(self) -> None:
        if not self.cfg.feedback:
            # Custom Interaction addresses don't answer /? queries — no listener.
            log.info("OSC out -> %s:%d (feedback disabled)", self.cfg.host, self.cfg.port)
            return
        disp = Dispatcher()
        disp.set_default_handler(self._on_feedback)
        try:
            self._server = ThreadingOSCUDPServer(("0.0.0.0", self.cfg.feedback_port), disp)
        except OSError as e:  # port taken: queries will time out, sends still work
            log.error("Cannot bind feedback port %d: %s — capture/test will fail",
                      self.cfg.feedback_port, e)
            return
        self._server_thread = threading.Thread(
            target=self._server.serve_forever, daemon=True, name="osc-feedback")
        self._server_thread.start()
        log.info("OSC out -> %s:%d, feedback listener on :%d",
                 self.cfg.host, self.cfg.port, self.cfg.feedback_port)

    def stop(self) -> None:
        if self._server:
            self._server.shutdown()
            self._server = None

    # --- absolute sends (engine tick) -----------------------------------------
    def send_value(self, address: str, value: float) -> None:
        """Send one float to an explicit OSC address. Addresses are configured
        per beamer (e.g. /front/scale/1, /front/positionV/1), so both custom
        Interaction bindings and the standard /layer:NAME API are just data."""
        if not address:
            return
        try:
            self._client.send_message(address, float(value))
        except OSError as e:
            log.warning("OSC send failed: %s", e)

    def send_bang(self, address: str) -> None:
        """Send an argument-less OSC message (pure path trigger, e.g. to reveal
        a layer in Millumin: the address alone, no value)."""
        if not address:
            return
        try:
            self._client.send_message(address, [])
        except OSError as e:
            log.warning("OSC send failed: %s", e)

    # --- correlated query (blocking; run via asyncio.to_thread) ---------------
    def query(self, layer: str, timeout_ms: Optional[int] = None) -> Optional[dict]:
        """Current {scale, pos_x, pos_y, latency_ms} of a layer, or None on
        timeout (Millumin absent / wrong layer name / feedback disabled —
        indistinguishable causes)."""
        timeout = (timeout_ms or self.cfg.feedback_timeout_ms) / 1000.0
        base = f"/millumin/layer:{layer}"
        slot = {"event": threading.Event(), "scale": None, "pos_x": None, "pos_y": None}
        addresses = (f"{base}/scale", f"{base}/position/xy",
                     f"{base}/position/x", f"{base}/position/y")
        with self._lock:
            for a in addresses:
                self._pending[a] = slot
        t0 = time.monotonic()
        try:
            self._client.send_message(f"/layer:{layer}/scale/?", [])
            self._client.send_message(f"/layer:{layer}/position/xy/?", [])
            deadline = t0 + timeout
            while time.monotonic() < deadline:
                if slot["event"].wait(timeout=deadline - time.monotonic()):
                    slot["event"].clear()
                    if all(slot[k] is not None for k in ("scale", "pos_x", "pos_y")):
                        break
        except OSError as e:
            log.warning("OSC query send failed: %s", e)
        finally:
            with self._lock:
                for a in addresses:
                    if self._pending.get(a) is slot:
                        del self._pending[a]
        if any(slot[k] is None for k in ("scale", "pos_x", "pos_y")):
            return None
        return {
            "scale": float(slot["scale"]),
            "pos_x": float(slot["pos_x"]),
            "pos_y": float(slot["pos_y"]),
            "latency_ms": round((time.monotonic() - t0) * 1000.0, 1),
        }

    def _on_feedback(self, address: str, *args) -> None:
        with self._lock:
            slot = self._pending.get(address)
        if slot is None or not args:
            return  # unsolicited feedback traffic — ignored
        try:
            if address.endswith("/scale"):
                slot["scale"] = float(args[0])
            elif address.endswith("/position/xy") and len(args) >= 2:
                slot["pos_x"], slot["pos_y"] = float(args[0]), float(args[1])
            elif address.endswith("/position/x"):
                slot["pos_x"] = float(args[0])
            elif address.endswith("/position/y"):
                slot["pos_y"] = float(args[0])
        except (TypeError, ValueError):
            return
        slot["event"].set()

File: cadreur-rs/src/millumin.rs
//! Millumin OSC output over UDP.
//!
//! Ported from `src/cadreur/millumin.py`, minus its feedback listener and
//! correlated `query()`. That path is unreachable in practice: `query` is
//! called from nowhere in the Python, `probe_enabled` is passed to the engine
//! and never stored, and `millumin.feedback` defaults to false because custom
//! Interaction addresses — the configuration actually in use — do not answer
//! `/?` readback at all. Carrying it would have meant porting a pending-reply
//! correlation table that nothing can reach.
//!
//! Addresses are configured per channel (`/front/scale/1`,
//! `/front/positionV/1`, ...), so both custom Interaction bindings and the
//! standard `/layer:NAME` API are just data.
//!
//! Sends are deliberately synchronous. UDP never meaningfully blocks, and
//! keeping them sync keeps the engine tick synchronous and testable with a
//! fake sender, exactly as the Python did.

use std::net::{SocketAddr, ToSocketAddrs, UdpSocket};

use rosc::{OscMessage, OscPacket, OscType, encoder};

use crate::config::MilluminCfg;

/// What the engine needs. A test double implements this to capture sends.
pub trait OscSender: Send + Sync {
    /// One float to an explicit address.
    fn send_value(&self, address: &str, value: f64);
    /// An argument-less message — a pure path trigger, used to reveal a layer.
    fn send_bang(&self, address: &str);
}

pub struct MilluminIo {
    socket: Option<UdpSocket>,
    dest: Option<SocketAddr>,
}

impl MilluminIo {
    pub fn new(cfg: &MilluminCfg) -> Self {
        let dest = (cfg.host.as_str(), cfg.port)
            .to_socket_addrs()
            .ok()
            .and_then(|mut it| it.next());
        let socket = match UdpSocket::bind("0.0.0.0:0") {
            Ok(s) => Some(s),
            Err(e) => {
                eprintln!("Cannot open the OSC socket: {e} — Millumin will not be driven");
                None
            }
        };
        match dest {
            Some(d) => eprintln!("OSC out -> {d}"),
            None => eprintln!("Cannot resolve {}:{} — Millumin will not be driven", cfg.host, cfg.port),
        }
        Self { socket, dest }
    }

    fn send(&self, address: &str, args: Vec<OscType>) {
        if address.is_empty() {
            return;
        }
        let (Some(socket), Some(dest)) = (&self.socket, self.dest) else { return };
        let packet = OscPacket::Message(OscMessage { addr: address.to_string(), args });
        let Ok(buf) = encoder::encode(&packet) else {
            eprintln!("Cannot encode OSC for {address}");
            return;
        };
        if let Err(e) = socket.send_to(&buf, dest) {
            eprintln!("OSC send failed: {e}");
        }
    }
}

impl OscSender for MilluminIo {
    fn send_value(&self, address: &str, value: f64) {
        // python-osc sends Python floats as OSC float32; keep the wire format
        // identical so Millumin sees no difference.
        self.send(address, vec![OscType::Float(value as f32)]);
    }

    fn send_bang(&self, address: &str) {
        self.send(address, vec![]);
    }
}

/// Captures sends so engine tests can assert on OSC traffic.
#[cfg(test)]
#[derive(Default)]
pub struct FakeSender {
    pub sent: std::sync::Mutex<Vec<(String, Option<f64>)>>,
}

#[cfg(test)]
impl FakeSender {
    /// Every value sent to one address, in order.
    pub fn values_to(&self, address: &str) -> Vec<f64> {
        self.sent
            .lock()
            .expect("lock")
            .iter()
            .filter(|(a, v)| a == address && v.is_some())
            .map(|(_, v)| v.expect("filtered"))
            .collect()
    }

    pub fn any_to(&self, addresses: &[&str]) -> bool {
        self.sent.lock().expect("lock").iter().any(|(a, _)| addresses.contains(&a.as_str()))
    }

    pub fn clear(&self) {
        self.sent.lock().expect("lock").clear();
    }

    pub fn is_empty(&self) -> bool {
        self.sent.lock().expect("lock").is_empty()
    }
}

#[cfg(test)]
impl OscSender for FakeSender {
    fn send_value(&self, address: &str, value: f64) {
        self.sent.lock().expect("lock").push((address.to_string(), Some(value)));
    }

    fn send_bang(&self, address: &str) {
        self.sent.lock().expect("lock").push((address.to_string(), None));
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn encodes_a_float_message_like_python_osc() {
        let packet = OscPacket::Message(OscMessage {
            addr: "/front/scale/1".into(),
            args: vec![OscType::Float(0.5)],
        });
        let buf = encoder::encode(&packet).expect("encodes");
        // address padded to a 4-byte boundary, then the ",f" type tag
        assert!(buf.starts_with(b"/front/scale/1\0\0"));
        assert!(buf.windows(2).any(|w| w == b",f"));
    }

    #[test]
    fn a_bang_carries_no_arguments() {
        let packet =
            OscPacket::Message(OscMessage { addr: "/front/layer/1".into(), args: vec![] });
        let buf = encoder::encode(&packet).expect("encodes");
        assert!(buf.windows(1).any(|w| w == b","));
        // no 'f' type tag: the trigger is the address alone
        let tags_start = buf.iter().position(|&b| b == b',').expect("type tag");
        assert_eq!(buf[tags_start + 1], 0, "expected an empty type tag list");
    }

    #[test]
    fn empty_address_is_ignored() {
        let io = MilluminIo { socket: None, dest: None };
        io.send_value("", 1.0); // must not panic
        io.send_bang("");
    }

    #[test]
    fn fake_sender_records_both_kinds() {
        let f = FakeSender::default();
        f.send_value("/a", 0.25);
        f.send_bang("/b");
        let sent = f.sent.lock().expect("lock");
        assert_eq!(sent[0], ("/a".to_string(), Some(0.25)));
        assert_eq!(sent[1], ("/b".to_string(), None));
    }
}
